### PR TITLE
[OPIK-397] Add spans stats endpoint

### PR DESCRIPTION
**Pull request by thiagohora**
## Details
Add spans stats endpoint

## Issues
OPIK-397

### CODE REVIEW

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java [hunk 1]
```diff
@@ -4758,6 +4772,2674 @@ void feedback__whenFeedbackSpanBatchHasMaxSize__thenReturnNoContentAndCreateScor
         }
     }
 
+    @Nested
+    @DisplayName("Get span stats:")
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    class GetSpanStats {
+
+        private List<Double> calculateQuantiles(List<Double> data, List<Double> quantiles) {
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]: Don't you prefer using an out-of-the-box solution? For example guice's [Quantiles](https://guava.dev/releases/23.0/api/docs/com/google/common/math/Quantiles.html)

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java [hunk 2]
```diff
@@ -4758,6 +4774,2671 @@ void feedback__whenFeedbackSpanBatchHasMaxSize__thenReturnNoContentAndCreateScor
...
+
+            batchCreateAndAssert(unexpectedSpans, apiKey, workspaceName);
+
+            List<ProjectStats.ProjectStatItem<?>> projectStatItems = getProjectStatItems(spans);
+
+            getStatsAndAssert(null, projectId, null, traceId, type, apiKey, workspaceName, projectStatItems);
+        }
+
+        @Test
+        void getTraceStats__whenFilterIdAndNameEqual__thenReturnSpansFiltered() {
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]:
```suggestion
        void getSpanStats__whenFilterIdAndNameEqual__thenReturnSpansFiltered() {
```

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java [hunk 3]
```diff
@@ -4758,6 +4774,2671 @@ void feedback__whenFeedbackSpanBatchHasMaxSize__thenReturnNoContentAndCreateScor
...
+                            .value(spans.getFirst().name())
+                            .build());
+
+            List<ProjectStats.ProjectStatItem<?>> projectStatItems = getProjectStatItems(expectedSpans);
+
+            getStatsAndAssert(projectName, null, filters, null, null, apiKey, workspaceName, projectStatItems);
+        }
+
+        @Test
+        void getTraceStats__whenFilterNameEqual__thenReturnSpansFiltered() {
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]:
```suggestion
        void getSpanStats__whenFilterNameEqual__thenReturnSpansFiltered() {
```

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java [hunk 4]
```diff
@@ -4758,6 +4774,2671 @@ void feedback__whenFeedbackSpanBatchHasMaxSize__thenReturnNoContentAndCreateScor
...
+                    .value(spans.getFirst().name().toUpperCase())
+                    .build());
+
+            List<ProjectStats.ProjectStatItem<?>> projectStatItems = getProjectStatItems(expectedSpans);
+
+            getStatsAndAssert(projectName, null, filters, null, null, apiKey, workspaceName, projectStatItems);
+        }
+
+        @Test
+        void getTraceStats__whenFilterNameStartsWith__thenReturnSpansFiltered() {
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]:
```suggestion
        void getSpanStats__whenFilterNameStartsWith__thenReturnSpansFiltered() {
```

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java [hunk 5]
```diff
@@ -4758,6 +4774,2671 @@ void feedback__whenFeedbackSpanBatchHasMaxSize__thenReturnNoContentAndCreateScor
...
+                    .value(spans.getFirst().name().substring(0, spans.getFirst().name().length() - 4).toUpperCase())
+                    .build());
+
+            List<ProjectStats.ProjectStatItem<?>> projectStatItems = getProjectStatItems(expectedSpans);
+
+            getStatsAndAssert(projectName, null, filters, null, null, apiKey, workspaceName, projectStatItems);
+        }
+
+        @Test
+        void getTraceStats__whenFilterNameEndsWith__thenReturnSpansFiltered() {
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]:
```suggestion
        void getSpanStats__whenFilterNameEndsWith__thenReturnSpansFiltered() {
```

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java (hunk 1 above)
Comment (thiagohora) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]: Great suggestion. I was not aware about that

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java:1]: Most of the tests should be renamed `.*Trace.*` to `.*Span.*`

**Review by idoberko2 — APPROVED** [review]
LGTM. Added a minor renaming comment to `SpansResourceTest`.

In addition, keep in mind that using PODAM for generation of feedback scores and usage tokens result in random keys in these objects. Therefore I'm not sure the aggregation is covered properly for these entities. Up to you whether to change it or not.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java [hunk 6]
```diff
@@ -4758,6 +4774,2671 @@ void feedback__whenFeedbackSpanBatchHasMaxSize__thenReturnNoContentAndCreateScor
...
+                    .stream()
+                    .sorted()
+                    .forEach(key -> stats
+                            .add(new ProjectStats.AvgValueStat("%s.%s".formatted("feedback_score", key),
+                                    feedback.get(key))));
+
+            return stats;
+        }
+
+        private Map<String, Double> calculateUsageAverage(List<Span> traces) {
```
Comment (BorisTkachenko) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]:
```suggestion
        private Map<String, Double> calculateUsageAverage(List<Span> spans) {
```

**Review by BorisTkachenko — COMMENTED** [review]
There are numerous typos in `SpansResourceTest.java` traces <-> spans, not only in test names, but in code as well. 
Also please take into consideration, that for spans usage you generate random data. But for use case, values could be only positive, not sure if it's relevant for your changes.

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java (hunk 6 above)
Comment (thiagohora) [reply apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]: Fixed

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java [hunk 7]
```diff
@@ -4822,6 +4842,2522 @@ void feedback__whenFeedbackSpanBatchHasMaxSize__thenReturnNoContentAndCreateScor
...
+    @Nested
+    @DisplayName("Get span stats:")
+    @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+    class GetSpanStats {
+
+        @Test
+        void findWithUsage() {
+            var projectName = RandomStringUtils.randomAlphanumeric(10);
+            var spans = PodamFactoryUtils.manufacturePojoList(podamFactory, Span.class).stream()
+                    .map(trace -> trace.toBuilder()
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]:
```suggestion
                    .map(span -> span.toBuilder()
```

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java [hunk 8]
```diff
@@ -4822,6 +4842,2522 @@ void feedback__whenFeedbackSpanBatchHasMaxSize__thenReturnNoContentAndCreateScor
...
+        @Test
+        void findWithoutUsage() {
+            var apiKey = UUID.randomUUID().toString();
+            var workspaceName = RandomStringUtils.randomAlphanumeric(10);
+            var workspaceId = UUID.randomUUID().toString();
+            mockTargetWorkspace(apiKey, workspaceName, workspaceId);
+
+            var projectName = RandomStringUtils.randomAlphanumeric(10);
+            var spans = PodamFactoryUtils.manufacturePojoList(podamFactory, Span.class).stream()
+                    .map(trace -> trace.toBuilder()
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]:
```suggestion
                    .map(span -> span.toBuilder()
```

**Review by idoberko2 — APPROVED** [review]
Two minor typos, other than that LGTM

**Review by idoberko2 — APPROVED** [review]

**Pull request merged**